### PR TITLE
Implemented versioning

### LINKED ISSUE #83: Versioning support
| Q | A
| --- | ---
| Bug? | no
| New Feature? | yes
| Sulu Version | `^1.5`
| Browser Version | none

#### Expected Behavior

Versioning should behave like sulu pages. For this we have to drop support for sulu `1.4`.

### PR DESCRIPTION
**Pull request by trickreich**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #83 
| License | MIT
| Documentation PR | --

#### What's in this PR?

Implemented versioning for articles.

#### Why?

Because it's really nice.

### CODE REVIEW

File: Controller/VersionController.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+use Sulu\Component\Rest\ListBuilder\ListRepresentation;
+use Sulu\Component\Rest\RequestParametersTrait;
+use Sulu\Component\Security\SecuredControllerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Handles the versions of articles.
+ */
+class VersionController extends FOSRestController implements
```
Comment (danrot) [inline Controller/VersionController.php]: @trickreich Can you please show which lines exactly you had to adapt? Then we can discuss what solution would be the best one.

**Comment by trickreich** [comment]
image

**Comment by trickreich** [comment]
@danrot @wachterjohannes I've uploaded the diff, now you guys can discuss this 👍

**Comment by wachterjohannes** [comment]
@trickreich we wont move more to sulu - so we have to implement it here like you have done in this PR. please add tests for this

**Comment by danrot** [comment]
Well, just because this change didn't make it into 1.5 doesn't mean we will change something about that into 1.6... Would still like to discuss that change for 1.6, maybe we find a nice solution. But that doesn't have to be now.

**Comment by trickreich** [comment]
But when I have do implement this now, we should discuss this now.
@danrot @wachterjohannes

**Comment by wachterjohannes** [comment]
@trickreich @danrot we have to copy this code now from sulu. because we cannot wait for sulu `1.6` release. maybe we create an issue in sulu and here which says that some code can be shared. so @trickreich go on implementing this feature (:

**Comment by danrot** [comment]
Creating an issue sounds like a good idea. Sorry that this change didn't make it into the 1.5 release...

File: Controller/TemplateController.php [hunk 2]
```diff
@@ -55,4 +56,21 @@ public function getAction(Request $request)
...
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return Response
+     */
+    public function settingsAction(Request $request)
+    {
+        return $this->render(
+            'SuluArticleBundle::settings.html.twig',
+            [
+                'versioning' => $this->getParameter('sulu_document_manager.versioning.enabled'),
```
Comment (trickreich) [inline Controller/TemplateController.php]: @wachterjohannes Should we create a own parameter or use it from the Sulu document manager?

Comment (wachterjohannes) [inline Controller/TemplateController.php]: i would use that. when i enable i want it everywhere.

File: .travis.yml [hunk 3]
```diff
@@ -18,10 +23,25 @@ matrix:
     - php: 7.0
       env:
       # - COMPOSER_FLAGS="--prefer-dist --no-interaction"
-        - SULU_VERSION="~1.5@RC"
+        - SULU_VERSION="~1.5"
+        - CODE_COVERAGE="--coverage-clover=coverage.clover"
```
Comment (wachterjohannes) [inline .travis.yml]: core coverage only in one testrun.

File: .travis.yml [hunk 4]
```diff
@@ -5,6 +5,11 @@ language: php
...
   - elasticsearch
 
+env:
+  global:
+    - JACKRABBIT_VERSION=2.12.0
+    - SYMFONY__PHPCR__TRANSPORT=doctrinedbal
+
 cache:
   directories:
     - "$HOME/.composer/cache"
```
Comment (wachterjohannes) [inline .travis.yml]: you can add the `downloads` directory here.

File: .travis.yml [hunk 5]
```diff
@@ -18,10 +23,25 @@ matrix:
...
+      env:
+        - SULU_VERSION="~1.5"
+        - SYMFONY__PHPCR__TRANSPORT=jackrabbit
         - CODE_COVERAGE="--coverage-clover=coverage.clover"
 
 before_install:
+  - if [ ! -d downloads ]; then mkdir downloads; fi
+  - |
+    if [[ $SYMFONY__PHPCR__TRANSPORT = jackrabbit ]]; then
+        if [ ! -f downloads/jackrabbit-standalone-$JACKRABBIT_VERSION.jar ]; then
```
Comment (wachterjohannes) [inline .travis.yml]: where is `JACKRABBIT_VERSION` defined?

File: Controller/VersionController.php [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+     */
+    public function cgetAction(Request $request, $uuid)
+    {
+        $locale = $this->getRequestParameter($request, 'locale', true);
+
+        $document = $this->get('sulu_document_manager.document_manager')->find($uuid, $request->query->get('locale'));
+        $versions = array_reverse(array_filter($document->getVersions(), function($version) use ($locale) {
+            /** @var Version $version */
+            return $version->getLocale() === $locale;
+        }));
```
Comment (wachterjohannes) [inline Controller/VersionController.php]: formating is here quite strange use an own line function call.

File: Controller/VersionController.php [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        $total = count($versions);
+
+        $listRestHelper = $this->get('sulu_core.list_rest_helper');
+        $limit = $listRestHelper->getLimit();
+
+        $versions = array_slice($versions, $listRestHelper->getOffset(), $limit);
+
+        $userIds = array_unique(array_map(function($version) {
```
Comment (wachterjohannes) [inline Controller/VersionController.php]: same here.

File: Resources/public/js/services/manager.js [hunk 8]
```diff
@@ -16,6 +16,14 @@ define(['jquery', 'services/husky/util'], function($, Util) {
             '<% if (typeof id !== "undefined") { %>/<%= id %><% } %>' +
             '?locale=<%= locale %>' +
             '<% if (typeof action !== "undefined") { %>&action=<%= action %><% } %>'
+        ),
+        urlVersion: _.template(
```
Comment (wachterjohannes) [inline Resources/public/js/services/manager.js]: could we do here something like:

```
urlTemplate = '/admin/api/articles<% if (typeof id !== "undefined") { %>/<%= id %><% } %><%= postfix %>?locale=<%= locale %><% if (typeof action !== "undefined") { %>&action=<%= action %><% } %>';
```

then we only have to manage one url template

File: composer.json [hunk 9]
```diff
@@ -5,7 +5,7 @@
     "license": "MIT",
     "require": {
         "php": "^5.5 || ^7.0",
-        "sulu/sulu": "^1.4 || dev-develop",
+        "sulu/sulu": "^1.4 || dev-de^velop",
```
Comment (wachterjohannes) [inline composer.json]: i think thats a typo (:

File: Tests/Functional/Controller/VersionControllerTest.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        if (!$this->getContainer()->getParameter('sulu_document_manager.versioning.enabled')) {
+            $this->markTestSkipped('Versioning is not enabled');
+        }
+
+        $this->initPhpcr();
+        $this->documentManager = $this->getContainer()->get('sulu_document_manager.document_manager');
+    }
+
+    public function testPostRestore()
```
Comment (wachterjohannes) [inline Tests/Functional/Controller/VersionControllerTest.php]: add also a testcase for a not existing version.

**Comment by wachterjohannes** [comment]
can you also remove the debug flag in .travis.yml in line https://github.com/sulu/SuluArticleBundle/pull/85/files#diff-354f30a63fb0907d4ad57269548329e3R59

File: .travis.yml (hunk 5 above)
Comment (trickreich) [inline .travis.yml]: line 10

**Pull request merged**